transport/tchannel/peer.go

```diff
@@ -0,0 +1,153 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package tchannel
+
+import (
+	"context"
+	"time"
+
+	"go.uber.org/yarpc/api/peer"
+	"go.uber.org/yarpc/peer/hostport"
+)
+
+type tchannelPeer struct {
+	*hostport.Peer
+	transport *Transport
+	addr      string
+	changed   chan struct{}
+	released  chan struct{}
+	timer     *time.Timer
+}
+
+func newPeer(pid hostport.PeerIdentifier, t *Transport) *tchannelPeer {
+	// Create a defused timer for later use.
+	timer := time.NewTimer(0)
+	if !timer.Stop() {
+		<-timer.C
+	}
+
+	return &tchannelPeer{
+		addr:      pid.Identifier(),
+		Peer:      hostport.NewPeer(pid, t),
+		transport: t,
+		changed:   make(chan struct{}, 1),
+		released:  make(chan struct{}, 0),
+		timer:     timer,
+	}
+}
+
+func (p *tchannelPeer) MaintainConn() {
+	cancel := func() {}
+
+	backoff := p.transport.connBackoffStrategy.Backoff()
+	var attempts uint
+
+	// Wait for start (so we can be certain that we have a channel).
+	<-p.transport.once.Started()
+	pl := p.transport.peerList()
+	if pl == nil {
+		return
+	}
+
+	// Attempt to retain an open connection to each peer so long as it is
+	// retained.
+	for {
```

Comment:
This may be gibberish, you know the whole picture better than I do, I'm guessing that something else breaks this? But from reading this, it seems like it has the potential to be an infinite loop - do you want some kind of way to break the loop in the worst case?